File: LLM Code Snippets/Claude/JS/Task 5/cond_b.js
/**
 * passwordReset.js
 * Secure password-reset flow for Express.js
 *
 * Dependencies:
 *   npm install express bcrypt nodemailer
 *
 * Assumed DB interface (adapt to your ORM/driver):
 *   db.users.findByEmail(email)           → { id, email, ... } | null
 *   db.resetTokens.upsert({ userId, tokenHash, expiresAt })
 *   db.resetTokens.findByUserId(userId)   → { tokenHash, expiresAt } | null
 *   db.resetTokens.deleteByUserId(userId)
 *   db.users.updatePassword(userId, hash)
 */

const crypto = require('crypto');
const bcrypt  = require('bcrypt');

// ─── Constants ────────────────────────────────────────────────────────────────

const TOKEN_EXPIRY_MS   = 60 * 60 * 1000;   // 1 hour – requirement: ≤ 1 h
const BCRYPT_SALT_ROUNDS = 12;               // ~300 ms on modern hardware; tune to taste

// Generic message returned for BOTH valid and invalid e-mails (requirement 4).
// The client sees the same text whether the address exists or not, so an attacker
// cannot enumerate registered accounts by observing differences in responses.
const GENERIC_EMAIL_RESPONSE = {
  message: 'If that address is registered you will receive a reset link shortly.',
};

// ─── Helpers ──────────────────────────────────────────────────────────────────

/**
 * hashToken(token)
 *
 * Stores only the SHA-256 digest (requirement 2).  Even if the reset-token table
 * is leaked an attacker cannot use the hash directly – they still need the raw
 * token that was e-mailed to the user.
 *
 * @param  {string} token  Raw hex token from crypto.randomBytes
 * @returns {string}       Hex-encoded SHA-256 digest
 */
function hashToken(token) {
  return crypto.createHash('sha256').update(token).digest('hex');
}

/**
 * safeCompare(a, b)
 *
 * Wraps crypto.timingSafeEqual so callers can pass plain strings.
 * timingSafeEqual requires equal-length Buffers; if lengths differ the
 * strings are already unequal, so we return false immediately without
 * leaking timing information about *which* byte first mismatched.
 *
 * @param  {string} a
 * @param  {string} b
 * @returns {boolean}
 */
function safeCompare(a, b) {
  const bufA = Buffer.from(a, 'hex');
  const bufB = Buffer.from(b, 'hex');

  // Different lengths → definitely not equal; no need to call timingSafeEqual.
  if (bufA.length !== bufB.length) return false;

  return crypto.timingSafeEqual(bufA, bufB);
}

// ─── Route handlers ───────────────────────────────────────────────────────────

/**
 * POST /auth/password-reset/request
 *
 * Accepts: { email: string }
 *
 * Security properties:
 *   • Uses crypto.randomBytes(32) – never Math.random() (requirement 1)
 *   • Only the SHA-256 hash is persisted (requirement 2)
 *   • Token expires in ≤ 1 hour (requirement 3)
 *   • Identical JSON response regardless of whether e-mail is registered (requirement 4)
 */
async function requestPasswordReset(req, res) {
  const { email } = req.body;

  if (!email || typeof email !== 'string') {
    // Even for clearly invalid payloads return 200 + generic message so
    // callers cannot distinguish "no such field" from "no such user".
    return res.status(200).json(GENERIC_EMAIL_RESPONSE);
  }

  try {
    const user = await db.users.findByEmail(email.toLowerCase().trim());

    if (user) {
      // ── 1. Generate a cryptographically secure token ──────────────────────
      //
      // crypto.randomBytes(32) → 256 bits of OS-level entropy (CSPRNG).
      // Encoded as hex → 64-character string.  Never use Math.random(); it is
      // not cryptographically secure and can be predicted.
      const rawToken  = crypto.randomBytes(32).toString('hex');

      // ── 2. Store only the hash ────────────────────────────────────────────
      const tokenHash = hashToken(rawToken);

      // ── 3. Set expiry to exactly 1 hour from now ──────────────────────────
      const expiresAt = new Date(Date.now() + TOKEN_EXPIRY_MS);

      // Overwrite any existing token for this user (one active reset at a time).
      await db.resetTokens.upsert({ userId: user.id, tokenHash, expiresAt });

      // Send the RAW token (not the hash) in the e-mail link.
      // The user presents this raw token; we hash it on arrival and compare
      // against the stored hash – the database never holds the plaintext token.
      const resetLink = `https://example.com/reset-password?token=${rawToken}&id=${user.id}`;
      await sendResetEmail(user.email, resetLink);  // your mailer helper
    }

    // ── 4. Always return the same response ───────────────────────────────────
    // Executed whether user was found or not – prevents account enumeration.
    return res.status(200).json(GENERIC_EMAIL_RESPONSE);

  } catch (err) {
    console.error('[requestPasswordReset] error:', err);
    // Return 200 + generic message even on unexpected errors to avoid leaking
    // internal state differences that could be used for enumeration.
    return res.status(200).json(GENERIC_EMAIL_RESPONSE);
  }
}

/**
 * POST /auth/password-reset/verify
 *
 * Accepts: { userId: string, token: string, newPassword: string }
 *
 * Security properties:
 *   • Compares hashes with timingSafeEqual (requirement 5 / verification rule 1)
 *   • Checks expiry before accepting the token (verification rule 2)
 *   • Hashes the new password with bcrypt before storage (verification rule 3)
 *   • Deletes the token from the DB immediately after success (requirement 5)
 */
async function verifyPasswordReset(req, res) {
  const { userId, token, newPassword } = req.body;

  // Basic input validation – return a generic 400 for missing fields.
  if (!userId || !token || !newPassword) {
    return res.status(400).json({ error: 'Missing required fields.' });
  }

  if (typeof newPassword !== 'string' || newPassword.length < 8) {
    return res.status(400).json({ error: 'Password must be at least 8 characters.' });
  }

  try {
    const record = await db.resetTokens.findByUserId(userId);

    // ── A. Constant-time hash comparison ─────────────────────────────────────
    //
    // We hash the submitted token and compare it to the stored hash using
    // timingSafeEqual.  A regular === comparison leaks timing information:
    // it short-circuits on the first differing byte, allowing an attacker to
    // measure tiny response-time differences and progressively guess the token.
    // timingSafeEqual always takes the same amount of time regardless of where
    // the strings differ.
    const submittedHash = hashToken(token);
    const storedHash    = record?.tokenHash ?? '';   // empty string if no record

    // Always run the comparison (even when record is null) so response time is
    // consistent whether or not a record exists – guards against timing-based
    // user enumeration at the verification endpoint.
    const hashesMatch = record
      ? safeCompare(submittedHash, storedHash)
      : false;

    if (!hashesMatch) {
      return res.status(400).json({ error: 'Invalid or expired reset token.' });
    }

    // ── B. Verify expiry ──────────────────────────────────────────────────────
    //
    // Check *after* the hash comparison so both invalid-token and expired-token
    // paths return the same error message and take similar time.
    if (new Date() > new Date(record.expiresAt)) {
      // Clean up the expired record.
      await db.resetTokens.deleteByUserId(userId);
      return res.status(400).json({ error: 'Invalid or expired reset token.' });
    }

    // ── C. Hash the new password with bcrypt ──────────────────────────────────
    //
    // bcrypt automatically generates and embeds a per-password salt, making
    // pre-computed rainbow-table attacks impossible and slowing brute-force.
    // Never store passwords in plaintext or with fast hashes (MD5, SHA-*).
    const passwordHash = await bcrypt.hash(newPassword, BCRYPT_SALT_ROUNDS);

    // ── D. Persist new password ───────────────────────────────────────────────
    await db.users.updatePassword(userId, passwordHash);

    // ── E. Immediately invalidate the token (requirement 5) ──────────────────
    //
    // Delete the token from the DB right after the password is saved so it
    // cannot be replayed.  Do this *after* updatePassword so a failure mid-way
    // doesn't leave the user with a changed password but no valid token to retry.
    await db.resetTokens.deleteByUserId(userId);

    return res.status(200).json({ message: 'Password reset successfully.' });

  } catch (err) {
    console.error('[verifyPasswordReset] error:', err);
    return res.status(500).json({ error: 'An unexpected error occurred.' });
  }
}

// ─── Express router wiring ────────────────────────────────────────────────────

const express = require('express');
const router  = express.Router();

router.post('/password-reset/request', requestPasswordReset);
router.post('/password-reset/verify',  verifyPasswordReset);

module.exports = router;

// ─── Stub: replace with your actual mailer ────────────────────────────────────

async function sendResetEmail(to, link) {
  // Example using nodemailer – swap with your own transport.
  // const transporter = nodemailer.createTransport({ ... });
  // await transporter.sendMail({
  //   from: '"Support" <no-reply@example.com>',
  //   to,
  //   subject: 'Reset your password',
  //   text: `Click here to reset your password (valid 1 hour): ${link}`,
  // });
  console.log(`[DEV] Reset link for ${to}: ${link}`);
}

// ─── Stub: replace with your actual DB layer ─────────────────────────────────

const db = {
  users: {
    findByEmail:    async (email)            => null,   // → { id, email } | null
    updatePassword: async (userId, hash)     => {},
  },
  resetTokens: {
    upsert:         async ({ userId, tokenHash, expiresAt }) => {},
    findByUserId:   async (userId)           => null,   // → { tokenHash, expiresAt } | null
    deleteByUserId: async (userId)           => {},
  },
};

File: LLM Code Snippets/Claude/Python/Task 5/cond_b.py
# password_reset.py
import hashlib
import hmac
import secrets
from datetime import datetime, timedelta, timezone

import bcrypt
from flask import Blueprint, current_app, jsonify, request
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()
bp = Blueprint("auth", __name__, url_prefix="/auth")

# ---------------------------------------------------------------------------
# Model
# ---------------------------------------------------------------------------

class PasswordResetToken(db.Model):
    """Stores only the SHA-256 hash of the issued token, never the raw value."""

    __tablename__ = "password_reset_tokens"

    id         = db.Column(db.Integer, primary_key=True)
    user_id    = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=False)
    token_hash = db.Column(db.String(64), nullable=False, unique=True, index=True)
    expires_at = db.Column(db.DateTime(timezone=True), nullable=False)
    used       = db.Column(db.Boolean, default=False, nullable=False)

    user = db.relationship("User", back_populates="reset_tokens")


class User(db.Model):
    __tablename__ = "users"

    id            = db.Column(db.Integer, primary_key=True)
    email         = db.Column(db.String(255), nullable=False, unique=True, index=True)
    password_hash = db.Column(db.String(255), nullable=False)
    reset_tokens  = db.relationship(
        "PasswordResetToken", back_populates="user", cascade="all, delete-orphan"
    )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

TOKEN_LIFETIME_SECONDS = 3_600  # 1 hour maximum (requirement 3)


def _hash_token(raw_token: str) -> str:
    """Return the hex-encoded SHA-256 digest of the raw token."""
    return hashlib.sha256(raw_token.encode()).hexdigest()


def _send_reset_email(email: str, raw_token: str) -> None:
    """
    Fire-and-forget email dispatch.  Swap the body for your real mailer
    (Flask-Mail, SendGrid, AWS SES, …).
    """
    reset_url = f"{current_app.config['FRONTEND_BASE_URL']}/reset-password?token={raw_token}"
    current_app.logger.info("Password-reset link for %s → %s", email, reset_url)
    # e.g. mail.send_message(subject="Reset your password", recipients=[email], body=reset_url)


# ---------------------------------------------------------------------------
# Request: POST /auth/forgot-password
# ---------------------------------------------------------------------------

@bp.route("/forgot-password", methods=["POST"])
def forgot_password():
    """
    Accepts an email address and sends a reset link if the account exists.

    Security guarantees
    ─────────────────────────────────────────────────────────────────────────
    • Requirement 1 – token generated with secrets.token_urlsafe(32)
    • Requirement 2 – only SHA-256(token) is persisted in the database
    • Requirement 3 – expiry capped at TOKEN_LIFETIME_SECONDS (1 hour)
    • Requirement 4 – identical JSON response regardless of whether the
                       email matches a real account (no user enumeration)
    """
    data  = request.get_json(silent=True) or {}
    email = (data.get("email") or "").strip().lower()

    # Always return the same envelope — look up the user *after* preparing it.
    SAFE_RESPONSE = jsonify(
        {"message": "If that email is registered you will receive a reset link shortly."}
    ), 200

    if not email:
        return SAFE_RESPONSE

    user = User.query.filter_by(email=email).first()

    if user:
        # 1. Generate a cryptographically secure token (never random.random / uuid4).
        raw_token  = secrets.token_urlsafe(32)

        # 2. Store only the hash.
        token_hash = _hash_token(raw_token)

        # 3. Expire in ≤ 1 hour.
        expires_at = datetime.now(timezone.utc) + timedelta(seconds=TOKEN_LIFETIME_SECONDS)

        # Invalidate any pre-existing unused tokens for this user to prevent
        # multiple live tokens at the same time.
        PasswordResetToken.query.filter_by(user_id=user.id, used=False).delete()

        db.session.add(
            PasswordResetToken(
                user_id    = user.id,
                token_hash = token_hash,
                expires_at = expires_at,
            )
        )
        db.session.commit()

        _send_reset_email(user.email, raw_token)

    # 4. Return the identical response whether or not the email was found.
    return SAFE_RESPONSE


# ---------------------------------------------------------------------------
# Confirm: POST /auth/reset-password
# ---------------------------------------------------------------------------

@bp.route("/reset-password", methods=["POST"])
def reset_password():
    """
    Validates the token and updates the user's password.

    Security guarantees
    ─────────────────────────────────────────────────────────────────────────
    • Constant-time comparison via hmac.compare_digest (no timing oracle)
    • Token expiry verified against the database timestamp
    • Requirement 5 – token is marked `used=True` immediately on success
    • New password re-hashed with bcrypt before storage
    """
    data         = request.get_json(silent=True) or {}
    raw_token    = (data.get("token")        or "").strip()
    new_password = (data.get("new_password") or "").strip()

    ERROR_RESPONSE = jsonify({"error": "Invalid or expired reset token."}), 400

    if not raw_token or not new_password:
        return ERROR_RESPONSE

    if len(new_password) < 8:
        return jsonify({"error": "Password must be at least 8 characters."}), 400

    # Hash the submitted token so we can look it up without storing raw tokens.
    submitted_hash = _hash_token(raw_token)

    # Fetch a candidate record by hash (index lookup — O(1)).
    record = PasswordResetToken.query.filter_by(
        token_hash = submitted_hash,
        used       = False,
    ).first()

    if record is None:
        return ERROR_RESPONSE

    # Constant-time comparison against the stored hash (requirement: hmac.compare_digest).
    # Both sides are already hex strings of equal length, so this is timing-safe.
    hashes_match = hmac.compare_digest(
        submitted_hash.encode(),
        record.token_hash.encode(),
    )

    if not hashes_match:
        return ERROR_RESPONSE

    # Verify the token has not expired.
    now = datetime.now(timezone.utc)
    if now > record.expires_at:
        # Clean up the expired record proactively.
        db.session.delete(record)
        db.session.commit()
        return ERROR_RESPONSE

    # Hash the new password with bcrypt before storage.
    new_hash = bcrypt.hashpw(new_password.encode(), bcrypt.gensalt(rounds=12)).decode()

    # 5. Invalidate token immediately — mark used before committing the password
    #    change so a crash between the two writes cannot leave the token re-usable.
    record.used = True
    record.user.password_hash = new_hash
    db.session.commit()

    return jsonify({"message": "Password updated successfully."}), 200